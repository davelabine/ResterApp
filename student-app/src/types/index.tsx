export interface StudentPhoto {
    bucketName: string;
    key: string;
}

export interface StudentData {
    skey: string;
    id: string;
    name: string;
    photo: {};
}

export interface StoreState {
    studentList: Array<StudentData>;
    filter: string;
}