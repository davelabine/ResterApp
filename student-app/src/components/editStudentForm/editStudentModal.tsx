import * as React from 'react';
import { Button, Modal } from 'react-bootstrap';
import { EditStudentFormBase } from './editStudentFormBase';
import { StudentData } from '../../types';

export interface EditStudentModalProps {
    title: string;
    submitButtonText: string;
    initialStudent: StudentData;
    show: boolean;
    onHide: () => void;
    onSubmit: (student: StudentData, photo?: File) => void;
}

export interface EditStudentModalState {
    student: StudentData;
    photo?: File;
}

export class EditStudentModal extends React.Component<EditStudentModalProps, EditStudentModalState> {
    
    constructor(props: EditStudentModalProps) {
        super(props);

        this.state = { student: new StudentData() };

        this.onStudentTextChange = this.onStudentTextChange.bind(this);
        this.onStudentTextValidate = this.onStudentTextValidate.bind(this);
        this.onStudentFileChange = this.onStudentFileChange.bind(this);
        this.getValidationState = this.getValidationState.bind(this);
        this.onSubmit = this.onSubmit.bind(this);
        this.validateForm = this.validateForm.bind(this);
    }

    public componentWillReceiveProps(nextProps: EditStudentModalProps) {
        if (this.props.show !== nextProps.show) {
          this.setState( {student: this.props.initialStudent, photo: undefined} );
        }
    }

    public onStudentTextChange(label: string, value: string): void {
        let student = {...this.state.student};
        student[label] = value;
        this.setState( {student: student});
    }

    public onStudentFileChange(file: File) {
        console.log(file);
        this.setState( {photo: file} );
    }

    public onStudentTextValidate(label: string): 'success' | 'warning' | 'error' {
        console.log('validating label - ' + label);

        let value = this.state.student[label];
        console.log('value is: ' + value);
        if ((value.length > 0) && (value.length < 100) ) { 
            // console.log('return success');
            return 'success'; 
        }
        console.log('return error');
        return 'error';
    }

    public getValidationState(): 'success' | 'warning' | 'error' {
        return this.validateForm();
    }

    public onSubmit() {
        if (this.validateForm() === 'error') {
            console.log('Form is not valid - block onSubmit()');
            return;
        }
        console.log('Submitting student: ' + JSON.stringify(this.state.student));
        this.props.onSubmit(this.state.student, this.state.photo);
    }

    public render() {    
      return (
          <div className="static-modal">
          <Modal
            show={this.props.show}
            onHide={this.props.onHide}
            dialogClassName="custom-modal"
          >
            <Modal.Header>
              <Modal.Title>{this.props.title}</Modal.Title>
            </Modal.Header>
      
            <Modal.Body>
              <EditStudentFormBase
                  student={this.state.student}
                  onFormTextChange={this.onStudentTextChange}
                  onFormFileChange={this.onStudentFileChange}
                  onStudentTextValidate={this.onStudentTextValidate}
              />
            </Modal.Body>
      
            <Modal.Footer>
              <Button id="close" onClick={this.props.onHide}>Close</Button>
              <Button id="submit" onClick={this.onSubmit} bsStyle="primary">
                {this.props.submitButtonText}
              </Button>
            </Modal.Footer>
      
          </Modal>
        </div>
      );
    }

    private validateForm(): 'success' | 'warning' | 'error' {
        if ( this.onStudentTextValidate('lastName') === 'success' &&
             this.onStudentTextValidate('lastName') === 'success' &&
             this.onStudentTextValidate('studentId') === 'success' ) {
            return 'success'; 
        }
        return 'error'; 
    }
}