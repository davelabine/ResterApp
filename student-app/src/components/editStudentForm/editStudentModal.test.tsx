import * as React from 'react';
import { mount, shallow } from 'enzyme';
import { EditStudentModal } from './editStudentModal';
import { StudentData } from '../../types';
import toJson from 'enzyme-to-json';

describe('EditStudentModal', () => {
    it('renders without crashing and matches the last snapshot', () => {
        const form = mount(
                        <EditStudentModal 
                            title="Add Student"
                            submitButtonText="Add Student"
                            initialStudent={new StudentData()}
                            show={false}
                            onHide={jest.fn()}
                            onSubmit={jest.fn()}
                        />);
        expect(toJson(form)).toMatchSnapshot();
    });

    it('handles form events and submit', () => {
        const onSubmit = jest.fn();
        const form = shallow(
            <EditStudentModal 
                title="Add Student"
                submitButtonText="Add Student"
                initialStudent={new StudentData()}
                show={false}
                onHide={jest.fn()}
                onSubmit={onSubmit}
            />);
        const inst = form.instance() as EditStudentModal;

        inst.onStudentTextChange('id', '123');
        inst.onStudentTextChange('name', 'abc');
        expect(inst.state.student).toEqual({id: '123', name: 'abc'});

        const file = new File(['editStudentModal'], 'editStudentModal.test.txt', {
            type: 'text/plain',
          });
        inst.onStudentFileChange(file);
        expect(inst.state.photo).toEqual(file);

        inst.onSubmit();
        expect(onSubmit).toHaveBeenCalled();
    });
});