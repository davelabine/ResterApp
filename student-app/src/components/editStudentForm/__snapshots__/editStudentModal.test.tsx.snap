// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditStudentModal renders without crashing and matches the last snapshot 1`] = `
<EditStudentModal
  initialStudent={StudentData {}}
  onHide={[Function]}
  onSubmit={[Function]}
  show={false}
  submitButtonText="Add Student"
  title="Add Student"
>
  <div
    className="static-modal"
  >
    <Modal
      animation={true}
      autoFocus={true}
      backdrop={true}
      bsClass="modal"
      dialogClassName="custom-modal"
      dialogComponentClass={[Function]}
      enforceFocus={true}
      keyboard={true}
      manager={
        ModalManager {
          "add": [Function],
          "containers": Array [],
          "data": Array [],
          "handleContainerOverflow": true,
          "hideSiblingNodes": true,
          "isTopModal": [Function],
          "modals": Array [],
          "remove": [Function],
        }
      }
      onHide={[Function]}
      renderBackdrop={[Function]}
      restoreFocus={true}
      show={false}
    >
      <Modal
        autoFocus={true}
        backdrop={true}
        backdropClassName="modal-backdrop"
        backdropTransitionTimeout={150}
        containerClassName="modal-open"
        dialogTransitionTimeout={300}
        enforceFocus={true}
        keyboard={true}
        manager={
          ModalManager {
            "add": [Function],
            "containers": Array [],
            "data": Array [],
            "handleContainerOverflow": true,
            "hideSiblingNodes": true,
            "isTopModal": [Function],
            "modals": Array [],
            "remove": [Function],
          }
        }
        onEntering={[Function]}
        onExited={[Function]}
        onHide={[Function]}
        renderBackdrop={[Function]}
        restoreFocus={true}
        show={false}
        transition={[Function]}
      />
    </Modal>
  </div>
</EditStudentModal>
`;
